### ISSUE TITLE
[BUG] helm-operation pod uses incorrect taint for RKE2 Control Plane nodes, blocking provisioning

### ISSUE THREAD

**Issue by KingJ** [rancher/rancher#46228]
**Rancher Server Setup**
- Rancher version: 2.8.5
- Installation option (Docker install/Helm Chart): Helm Chart
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc): Rancher runs on EKS, issue is with a downstream RKE2 cluster.
- Proxy/Cert Details:

**Information about the Cluster**
- Kubernetes version: v1.27.14-rke2r1
- Cluster Type (Local/Downstream): Downstream
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider): Infrastructure Provider, OpenStack

**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom)
  - If custom, define the set of permissions: Admin

**Describe the bug**
As part of cluster provisioning, Rancher installs the `rancher-webhook` chart on to the downstream cluster. This appears to be a critical dependency which is performed after the etcd/control plane nodes have been bootstrapped, but before any worker nodes are allowed to be brought in - i.e. if the install operation for `rancher-webhook` fails, no workers will be brought in.

A `helm-operation` pod is spawned to install the webhook chart. This pod has the following tolerations;

```
Tolerations:     cattle.io/os=linux:NoSchedule
                 node-role.kubernetes.io/controlplane=true:NoSchedule
                 node-role.kubernetes.io/etcd=true:NoExecute
                 node.cloudprovider.kubernetes.io/uninitialized=true:NoSchedule
                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
```

While these tolerations are correct for RKE1, they are not correct for RKE2. Specifically;

* Control Plane nodes have a hyphen in `control-plane` and no value, i.e. `node-role.kubernetes.io/control-plane:NoSchedule`.
* etcd nodes have no value, i.e. `node-role.kubernetes.io/etcd:NoExecute`.

As a result, if provisioning an RKE2 cluster with dedicated control plane, etcd and worker pools the cluster will fail to initialise as the `helm-operation` pod cannot be scheduled on the control plane or etcd nodes (since the tolerations don't match the taints) and it can't run on the worker nodes either (as provisioning the webhook blocks the next step of cluster provisioning, i.e. the control plane being considered ready and thus bringing in worker nodes).

These tolerations are hardcoded [here](https://github.com/rancher/rancher/blob/7df10c297f33aa92ce57d4e3c162aea44418b9b3/pkg/catalogv2/helmop/operation.go#L1014-L1039). As far as I can tell, there is no way that I, as an end user, can override them.

To fix this issue, while maintaining RKE1 compatibility, I suggest that the following changes should be made;

* Add an additional `NoSchedule` toleration for `node-role.kubernetes.io/control-plane` (i.e. with the hyphen), but with no value and an operator of exists (as RKE2 does not set a value for this taint).
* Modify the existing `NoExecute` toleration for `node-role.kubernetes.io/etcd` to have no value and an operator of exists (which will still match the "true" value on RKE1, as well as the no-value in RKE2)

**To Reproduce**
Provision an RKE2 cluster, on any infrastructure provider, with dedicated Control Plane, etcd and Worker Node pools.

**Result**

**Expected Result**
The `helm-operation` pod starts successfully, and thus the cluster provisions successfully.

**Screenshots**
N/A

**Additional context**
* [A comment was left on the original commit that hardcoded the tolerations, noting that this does not work for RKE2](https://github.com/rancher/rancher/commit/90b90ba4ed9efe87cb4988b148b2b15a155e6b16#commitcomment-116143939).
* [Discussions on the implications of making RKE1 align to the Kubernetes standard for this taint](https://github.com/rancher/rancher/issues/41053#issuecomment-1736107500).

**Comment by dmpe** [comment]
Yes, we are also started to observe that. A fresh Rancher 2.8.5 cluster and entirely new rke2 1.28.10. And it was not like that... Dont known where that comes from. 

Did anyone found some solution to this ?

**Comment by KingJ** [comment]
> Did anyone found some solution to this ?

Not a solution, but a workaround - i'm setting the control plane nodepool to also have the worker role, then adding a custom `PreferNoSchedule` taint on it. This allows the `helm-operation` pods to schedule on it, and unblock provisioning.

It's not ideal though, as it means other pods can still land on the control plane nodes if there are no other available nodes. I want to keep the control plane as free as possible from non-control plane pods so this is a bit of a pain still.

**Comment by wotsyula** [comment]
Temporary workaround:

```yaml
---
apiVersion: provisioning.cattle.io/v1
kind: Cluster
metadata:
  labels:
  name: ${CLUSTER_NAME}
  namespace: fleet-default
spec:
  ...
  clusterAgentDeploymentCustomization:
    appendTolerations:
      - key: CriticalAddonsOnly
        operator: Exists
  fleetAgentDeploymentCustomization:
    appendTolerations:
      - key: CriticalAddonsOnly
        operator: Exists
```

I removed tolerations for control plane and ETCD since they were fixed in #48538

**Comment by brandond** [comment]
Note that "control-plane" is the correct form as per upstream docs:
https://kubernetes.io/docs/reference/labels-annotations-taints/#node-role-kubernetes-io-control-plane

**Comment by rasponien** [comment]
Hi. Any updates on this?
I'm having the same problem when importing AKS cluster with private API endpoint. I need to have the following tolerations block.
```
tolerations:
- key: "CriticalAddonsOnly"
  operator: "Exists"
``` 

My workaround was to this:

```
kubectl get pod helm-operation-xxx -n cattle-system -o yaml > helm-operation-pod.yaml
 
## change the name of the helm-operation
## change helm-operation pod tolerations block
 
kubectl apply -f helm-operation-pod.yaml
```

then setup Rancher webhook:

```
kubectl get deployment rancher-webhook -n cattle-system -o yaml > rancher-webhook-deployment.yaml
 
## Update tolerations
 
kubectl apply -f rancher-webhook-deployment.yaml
```

Rancher webhook then gets correctly set up, but even then I get multiple helm-operation pods in Pending state.

**Comment by diogoasouza** [comment]
Hi @rasponien, if I got this correctly, you are trying to add a custom toleration to a helm operation pod, correct? 
This issue is about RKE2 taints/tolerations, which should have been solved by https://github.com/rancher/rancher/pull/48538. 

That said, regarding your question, Rancher (v2.10+) does not currently support adding custom tolerations to helm-operation pods through the UI, [only](https://github.com/rancher/rancher/pull/46347) through the API. In regard to `rancher-webhook`, Rancher v2.10+ [automatically](https://github.com/rancher/rancher/pull/47592) adds tolerations to all taints present in CP nodes by default to helm operation pods used to install it, so the helm operation pod used to install it should be able to run in at least one node

**Comment by hansbogert** [comment]
@diogoasouza 

> Rancher v2.10+ https://github.com/rancher/rancher/pull/47592 adds tolerations to all taints present in CP nodes by default to helm operation pods used to install it, so the helm operation pod used to install it should be able to run in at least one node

That reasoning fails in case of AKS where there aren't any CP nodes.

**Comment by keisari-ch** [comment]
I second previous component.

AKS system pool node may have a CriticalAddonsOnly taint.
It is fine for cattle agents when using the append_tolerations mechanic.

But webhook and helm operations stay pending, some of our cluster sometimes have taint on all nodes.
